forward/testing/packet_synthesis.lua

```diff
@@ -0,0 +1,168 @@
+   local dividend = payload_len + fragment_len - 1
+   local num_fragments = (dividend - dividend % fragment_len) / fragment_len
+   local fragments = {n = num_fragments}
+   for i=1,num_fragments do
+      local curr_fragment_len = fragment_len
+      if i == num_fragments then
+         curr_fragment_len = payload_len - fragment_len * (num_fragments-1)
+      end
+      local fragment = string.sub(payload, (i-1) * fragment_len + 1, math.min(i * fragment_len, payload_len))
+      fragments[i] = fragment
+   end
+   return fragments
+end
+
+-- Arguments:
+-- src_mac, dst_mac (binary) -- source and destination MAC addresses;
+--                              should be the addresses of the router
+--                              (or last switch) and spike respectively
+-- src_addr, dst_addr (binary) -- source and destination IP addresses;
+--                                should be the addresses of the router
+--                                and spike respectively
+-- payload (binary) -- packet payload, defaults to 100 bytes of
+--                     generated rubbish
+-- payload_length (int, default 100) -- length of payload
+-- ip_flags (int, default 0) -- flags field for IP header
+-- frag_off (int, default 0) -- fragment offset field
+-- ttl (int, default 30) -- TTL field
+-- skip_tcp_header (bool, default nil) -- don't include a tcp header
+-- add_ip_gre_layer (bool, default nil) -- add a IP-GRE layer to test
+--                                       secondary fragmentation processing
+function make_ipv4_packet(config)
+   local payload_length = config.payload_length or 100
+   local payload = config.payload or
+      make_payload(payload_length)
+
+   local datagram = Datagram:new(nil, nil, {delayed_commit = true})
+   datagram:payload(payload, payload_length)
+   
+   local tcp_header_size = 0
+   if not config.skip_tcp_header then
+      local tcp_header = TCP:new({
+      })
+      tcp_header_size = tcp_header:sizeof()
+      -- TCP data offset field is in units of 32-bit words
+      tcp_header:offset(tcp_header_size / 4)
+      tcp_header:checksum()
+      datagram:push(tcp_header)
+   end
+
+   local ttl = config.ttl or 30
+   local ip_header = IPV4:new({
+      src = config.src_addr,
+      dst = config.dst_addr,
+      protocol = L4_TCP,
+      flags = config.ip_flags or 0,
+      frag_off = config.frag_off or 0,
+      ttl = ttl
+   })
+   ip_header:total_length(ip_header:sizeof() + tcp_header_size + payload_length)
+   ip_header:checksum()
+   datagram:push(ip_header)
+
+   if config.add_ip_gre_layer then
+      local gre_header = GRE:new({ protocol = L3_IPV4 })
+      datagram:push(gre_header)
+      local outer_ip_header = IPV4:new({
+         src = config.src_addr, -- Should be another spike's address (which should be equal to the inner IP header's dst address), just reusing router's address for now since it doesn't impact testing.
```

Comment:
I'm not sure I understand this comment - for instance the inner IP header should never have a spike's internal IP address.